pkgs/build-support/substitute/substitute.nix

```diff
@@ -6,7 +6,10 @@ args:
 # The `replacements` attribute should be a list of list of arguments
 # to `substitute`, such as `[ "--replace" "sourcetext" "replacementtext" ]`
 stdenvNoCC.mkDerivation ({
-  name = if args ? name then args.name else baseNameOf (toString args.src);
+  # unsafeDiscardStringContext is required here, because calling `lib.getName`
+  # on the resulting derivation gives the following error:
+  # "the string '50-coredump.conf' is not allowed to refer to a store path (such as '!out!/nix/store/zgf4bx2jrw6ysjlr3qf4i942z64d6559-systemd-253.5.drv')"
+  name = if args ? name then args.name else (builtins.unsafeDiscardStringContext (builtins.baseNameOf (toString args.src)));
```

Comment:
Even with the change you made to `getName`?

Reply:
Actually, I haven't tried. On the one hand it should work without, on the other hand only when using `getName`. Accessing `name` directly in other contexts would yield the same problem again. (So at the very least the comment mentioning `lib.getName` is now wrong)

Ideally, we'd fix all `name` attributes instead of `getName` IMO, but that's a big round of whack-a-mole to play

Reply:
One way to fix this once and for all would be to extend the idea of https://github.com/NixOS/nix/pull/7260 to `baseNameOf` and pretty much every function that outputs a string-with-context (or do this in `forceStringNoCtx`, which actually seems like a better idea to me now), but I'm not going to push this.

Reply:
Okay, so turns out to fix the problem at the generation site I'd have to fix every builder which might generate a default name from other parameters, or I could go one level deeper and add the `unsafeDiscardStringContext` somewhere in `mkDerivation`. I have no idea to which extend this is safe or not.

The same thing applies for my workaround adding `builtins.unsafeDiscardStringContext` to `lib.getName` (CC @infinisil).

I'm currently thinking about removing all my commits regarding this again, and instead simply call `lib.getName (builtins.unsafeDiscardStringContext pkgs)` in my code where needed. There I can at least guarantee that doing so is safe.